src/ImageSharp/Processing/Processors/Binarization/AdaptiveThresholdProcessor.cs

```diff
@@ -0,0 +1,115 @@
+﻿using System;
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Processors
+{
+    /// <summary>
+    /// Performs Bradley Adaptive Threshold filter against an image
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format of the image</typeparam>
+    internal class AdaptiveThresholdProcessor<TPixel> : IImageProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveThresholdProcessor{TPixel}"/> class.
+        /// </summary>
+        public AdaptiveThresholdProcessor()
+            : this(NamedColors<TPixel>.White, NamedColors<TPixel>.Black)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveThresholdProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="upper">Color for upper threshold</param>
+        /// <param name="lower">Color for lower threshold</param>
+        public AdaptiveThresholdProcessor(TPixel upper, TPixel lower)
+        {
+            this.Upper = upper;
+            this.Lower = lower;
+        }
+
+        /// <summary>
+        /// Gets or sets upper color limit for thresholding
+        /// </summary>
+        public TPixel Upper { get; set; }
+
+        /// <summary>
+        /// Gets or sets lower color limit for threshold
+        /// </summary>
+        public TPixel Lower { get; set; }
+
+        public unsafe void Apply(Image<TPixel> source, Rectangle sourceRectangle)
```

Comment:
Why `unsafe`?

Reply:
Oh! forgot to remove that, was doing some experimenting but since there are 2DArray pool I didn't implement it. Will remove it asap.